### PR TITLE
Add experimental feature DeferSendableChecking

### PR DESCRIPTION
**Pull request by JTurcotti**
This PR begins a project to add more sophisticated Sensibility checking to Swift's Concurrency model. In particular, sending non-`Sendable` values between differing isolations should only produce a diagnostic if it is possible that those values are accessible from, or permit access to, other values that would remain in scope after the send. This PR is the first step in that process. It adds an experimental feature flag `DeferSendableChecking` that, when enabled, prevents the `Sema` AST passes from emitting diagnostics about non-`Sendable` arguments passed to isolation-crossing calls. Instead, a SIL pass is added that checks `ApplyInst` instances in emitted SIL to see if their corresponding `ApplyExpr`s would've produced diagnostics. This pass will be the vehicle for more sophisticated checking in future PRs.

### CODE REVIEW

**Comment by JTurcotti** [comment]
@swift-ci please smoke test

File: include/swift/AST/Expr.h
Comment (DougGregor) [inline include/swift/AST/Expr.h:4711]: I wonder if we can come up with a better name here. `crossesIsolationDomain`? Perhaps even package this and the functions that follow it into a single function that returns an optional value that gives all of this information, e.g.,

```
llvm::Optional<IsolationCrossing> crossesIsolationDomain() const { ... }
```

where `IsolationCrossing` is some struct that has the caller and callee isolation domains. Maybe later we decide to move information about implicit async and such into it.

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp
Comment (DougGregor) [inline lib/SILOptimizer/Mandatory/SendNonSendable.cpp:19]: Can this feature check move way out? Either don't add the pass at all, or early-exit from the pass, if it's not set?

File: lib/Sema/TypeCheckConcurrency.cpp
Comment (DougGregor) [inline lib/Sema/TypeCheckConcurrency.cpp:2927]: This is all excellent, thank you!

File: test/Concurrency/DeferredSendableChecking/defer_and_typecheck_only.swift
Comment (DougGregor) [inline test/Concurrency/DeferredSendableChecking/defer_and_typecheck_only.swift:34]: This is a very nice approach!

**Review by DougGregor — COMMENTED** [review]
I like this approach a lot! A couple of comments about API design

File: include/swift/AST/Expr.h
Comment (kavon) [inline include/swift/AST/Expr.h:326]: Whenever you make a change to these bitfields, I believe you might have to increment the `SWIFTMODULE_VERSION_MINOR` counter at the top of `ModuleFormat.h`. Though maybe we don't serialize Expr's? I'd have to check

File: include/swift/AST/Expr.h
Comment (kavon) [inline include/swift/AST/Expr.h:4720]: I'm surprised this isn't a callee != caller isolation check. What if both are isolated but to different actors?

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp
Comment (kavon) [inline lib/SILOptimizer/Mandatory/SendNonSendable.cpp:21]: For efficiency, you need to check if this feature is enabled before looping over all of the instructions. You want to exit this method early if the feature isn't enabled. Otherwise this pass runs all the time, contributing to compilation time when not needed. The context of the instruction is the context of the function.

This fix in particular is definitely required before merging.

File: lib/Sema/TypeCheckConcurrency.cpp
Comment (kavon) [inline lib/Sema/TypeCheckConcurrency.cpp:1815]: I think we don't need to store the `DeclContext*` in the ApplyExpr.  Instead, take the `DeclContext*` as an argument here. Then, call `SILFunction::getDeclContext()` to retrieve the one corresponding to the instruction associated with an ApplyExpr.

**Review by kavon — CHANGES_REQUESTED** [review]

**Comment by kavon** [comment]
Overall I like this a lot more than the previous iteration!

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp
Comment (kavon) [reply lib/SILOptimizer/Mandatory/SendNonSendable.cpp:1]: ehh I spotted `#include "../AST/InlinableText.h"` in `TypeCheckMacros` so maybe it's OK. will defer to @DougGregor on this.

File: lib/Sema/TypeCheckConcurrency.cpp
Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.cpp:1815]: This is what I was trying to get at with my earlier comments, thanks Kavon

File: include/swift/AST/Expr.h
Comment (DougGregor) [reply include/swift/AST/Expr.h:326]: This isn't necessary, because we don't serialize `Expr` nodes.

File: test/Concurrency/DeferredSendableChecking/defer_and_typecheck_only.swift
Comment (DougGregor) [reply test/Concurrency/DeferredSendableChecking/defer_and_typecheck_only.swift:34]: Personally, I liked the documentation of deferred-warning, and was starting to wonder if we shouldn't do something sneaky where the input file has `deferred-warning`, and we run `sed` on it to replace that with `expected-warning` when doing the SIL-based diagnostics. That way, we don't have two files to keep in sync. (Not a requirement for merging, of course)

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp
Comment (DougGregor) [reply lib/SILOptimizer/Mandatory/SendNonSendable.cpp:1]: I probably shouldn't have done it there, either. I agree that we should move that function out to a separate header. Again, it can be a follow-up PR.

File: test/Concurrency/DeferredSendableChecking/defer_and_typecheck_only.swift
Comment (kavon) [reply test/Concurrency/DeferredSendableChecking/defer_and_typecheck_only.swift:34]: oh! that’s a good that idea for a future PR. so never mind let’s leave this as-is then.

**Comment by JTurcotti** [comment]
@swift-ci please smoke test

File: include/swift/AST/Expr.h
Comment (kavon) [inline include/swift/AST/Expr.h:4674]: This worked out really cleanly!

**Review by kavon — APPROVED** [review]

File: include/swift/AST/Expr.h
Comment (JTurcotti) [reply include/swift/AST/Expr.h:4674]: thanks!

**Pull request merged**
